### PR TITLE
null check a token's parent to prevent a NullReferenceException

### LINKED ISSUE #1187: Random crash (NullReferenceException) experienced while using signature help
I don't have any repro, just the crash information. The dump is available on internal bug 1127440.

    > Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.ErrorReporting.FatalError.Report(System.Exception exception, System.Action<System.Exception> handler) Unknown
      Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.SignatureHelp.CommonSignatureHelpUtilities.TryGetSyntax<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax>(Microsoft.CodeAnalysis.SyntaxNode root, int position, Microsoft.CodeAnalysis.LanguageServices.ISyntaxFactsService syntaxFacts, Microsoft.CodeAnalysis.Editor.SignatureHelpTriggerReason triggerReason, System.Func<Microsoft.CodeAnalysis.SyntaxToken, bool> isTriggerToken, System.Func<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax, Microsoft.CodeAnalysis.SyntaxToken, bool> isArgumentListToken, System.Threading.CancellationToken cancellationToken, out Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax expression) Unknown
      [Native to Managed Transition] 
      [Managed to Native Transition] 
      Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.SignatureHelp.CommonSignatureHelpUtilities.TryGetSyntax<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax>(Microsoft.CodeAnalysis.SyntaxNode root, int position, Microsoft.CodeAnalysis.LanguageServices.ISyntaxFactsService syntaxFacts, Microsoft.CodeAnalysis.Editor.SignatureHelpTriggerReason triggerReason, System.Func<Microsoft.CodeAnalysis.SyntaxToken, bool> isTriggerToken, System.Func<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax, Microsoft.CodeAnalysis.SyntaxToken, bool> isArgumentListToken, System.Threading.CancellationToken cancellationToken, out Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax expression) Unknown
      Microsoft.CodeAnalysis.CSharp.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.CSharp.SignatureHelp.AttributeSignatureHelpProvider.TryGetAttributeExpression(Microsoft.CodeAnalysis.SyntaxNode root, int position, Microsoft.CodeAnalysis.LanguageServices.ISyntaxFactsService syntaxFacts, Microsoft.CodeAnalysis.Editor.SignatureHelpTriggerReason triggerReason, System.Threading.CancellationToken cancellationToken, out Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax attribute) Unknown
      Microsoft.CodeAnalysis.CSharp.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.CSharp.SignatureHelp.AttributeSignatureHelpProvider.GetItemsWorkerAsync(Microsoft.CodeAnalysis.Document document, int position, Microsoft.CodeAnalysis.Editor.SignatureHelpTriggerInfo triggerInfo, System.Threading.CancellationToken cancellationToken) Unknown
      mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.__Canon>.Start<Microsoft.CodeAnalysis.Editor.CSharp.SignatureHelp.AttributeSignatureHelpProvider.<GetItemsWorkerAsync>d__5>(ref Microsoft.CodeAnalysis.Editor.CSharp.SignatureHelp.AttributeSignatureHelpProvider.<GetItemsWorkerAsync>d__5 stateMachine) Unknown
      Microsoft.CodeAnalysis.CSharp.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.CSharp.SignatureHelp.AttributeSignatureHelpProvider.GetItemsWorkerAsync(Microsoft.CodeAnalysis.Document document, int position, Microsoft.CodeAnalysis.Editor.SignatureHelpTriggerInfo triggerInfo, System.Threading.CancellationToken cancellationToken) Unknown
      Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.AbstractSignatureHelpProvider.GetItemsAsync(Microsoft.CodeAnalysis.Document document, int position, Microsoft.CodeAnalysis.Editor.SignatureHelpTriggerInfo triggerInfo, System.Threading.CancellationToken cancellationToken) Unknown
      mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.__Canon>.Start<Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.AbstractSignatureHelpProvider.<GetItemsAsync>d__15>(ref Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.AbstractSignatureHelpProvider.<GetItemsAsync>d__15 stateMachine) Unknown
      Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.AbstractSignatureHelpProvider.GetItemsAsync(Microsoft.CodeAnalysis.Document document, int position, Microsoft.CodeAnalysis.Editor.SignatureHelpTriggerInfo triggerInfo, System.Threading.CancellationToken cancellationToken) Unknown
      Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller.Session.ComputeItemsAsync(System.Collections.Generic.IList<Microsoft.CodeAnalysis.Editor.ISignatureHelpProvider> providers, Microsoft.VisualStudio.Text.SnapshotPoint caretPosition, Microsoft.CodeAnalysis.Editor.SignatureHelpTriggerInfo triggerInfo, Microsoft.CodeAnalysis.Document document, System.Threading.CancellationToken cancellationToken) Unknown
      mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.__Canon>.Start<Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller.Session.<ComputeItemsAsync>d__10>(ref Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller.Session.<ComputeItemsAsync>d__10 stateMachine) Unknown
      Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller.Session.ComputeItemsAsync(System.Collections.Generic.IList<Microsoft.CodeAnalysis.Editor.ISignatureHelpProvider> providers, Microsoft.VisualStudio.Text.SnapshotPoint caretPosition, Microsoft.CodeAnalysis.Editor.SignatureHelpTriggerInfo triggerInfo, Microsoft.CodeAnalysis.Document document, System.Threading.CancellationToken cancellationToken) Unknown
      Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Controller.Session.ComputeModelInBackgroundAsync(Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.SignatureHelp.Model currentModel, System.Collections.Generic.IList<Microsoft.CodeAnalysis.Editor.ISignatureHelpProvider> matchedProviders, System.Collections.Generic.IList<Microsoft.CodeAnalysis.Editor.ISignatureHelpProvider> unmatchedProviders, Microsoft.VisualStudio.Text.SnapshotPoint caretPosition, Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.DisconnectedBufferGraph disconnectedBufferGraph, Microsoft.CodeAnalysis.Editor.SignatureHelpTriggerInfo triggerInfo, System.Threading.CancellationToken cancellationToken) Unknown

### PR DESCRIPTION
**Pull request by brettfo**
This enables us to remove a try/catch block that was specifically added to find this NRE.

Fixes #1187.

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/CommonSignatureHelpUtilities.cs [hunk 1]
```diff
@@ -113,46 +113,36 @@ internal static class CommonSignatureHelpUtilities
...
-                        return true;
-                    }
+                    expression = token.GetAncestor<TSyntax>();
+                    return true;
                 }
-                else if (triggerReason == SignatureHelpTriggerReason.InvokeSignatureHelpCommand)
+            }
+            else if (triggerReason == SignatureHelpTriggerReason.InvokeSignatureHelpCommand)
+            {
+                expression = token.Parent?.GetAncestorsOrThis<TSyntax>().SkipWhile(syntax => !isArgumentListToken(syntax, token)).FirstOrDefault();
```
Comment (brettfo) [inline src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/CommonSignatureHelpUtilities.cs]: The true diff is adding the conditional access on this line `token.Parent?.GetAncestorsOrThis` and on one line below.

File: src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/CommonSignatureHelpUtilities.cs [hunk 2]
```diff
@@ -113,46 +113,36 @@ internal static class CommonSignatureHelpUtilities
...
+                return expression != null;
+            }
+            else if (triggerReason == SignatureHelpTriggerReason.RetriggerCommand)
+            {
+                if (!syntaxFacts.IsInNonUserCode(root.SyntaxTree, position, cancellationToken) ||
+                    syntaxFacts.IsEntirelyWithinStringOrCharLiteral(root.SyntaxTree, position, cancellationToken))
                 {
-                    expression = token.Parent.GetAncestorsOrThis<TSyntax>().SkipWhile(syntax => !isArgumentListToken(syntax, token)).FirstOrDefault();
+                    expression = token.Parent?.AncestorsAndSelf()
```
Comment (brettfo) [inline src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/CommonSignatureHelpUtilities.cs]: The true diff is adding the conditional access on this line `token.Parent?.AncestorsAndSelf` and on one line above.

**Comment by balajikris** [comment]
:+1:

**Comment by basoundr** [comment]
:+1:

**Comment by VladimirReshetnikov** [comment]
:+1:

**Comment by Pilchie** [comment]
Okay, now we know that we are getting asked to trigger signature help on a token with a null parent.  Where did that token come from?

**Comment by brettfo** [comment]
My WinDbg-fu is weak, but from what I can tell from the stack objects here's what I was able to re-create:

1. The C# AttributeSignatureHelpProvider got invoked inside the first UsingDirectiveSyntax in a file, presumably because a '(' or ',' was typed (one of the trigger characters.)
2. This led to AttributeSignatureHelpProvider::GetItemsWorker() -> TryGetAttributeExpression -> CommonSignatureHelpUtilities::TryGetSyntax().
3. ISyntaxFactsService.FindTokenOnLeftOfPosition() defaulted out and returned a CompilationUnitSyntax (see http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/Extensions/SyntaxNodeExtensions.cs,802) which has no parent and lead to the NRE.

The only way I could see this happening was if the trigger character was the first non-trivia token in the file, but I could never find a manual repro.

**Comment by Pilchie** [comment]
I wonder if this happens when the token is in skipped trivia because of some previous error.

**Comment by brettfo** [comment]
Digging deeper into this with @Pilchie's help we determined that the issue was not a CompilationUnitSyntax that didn't have a parent, but rather `syntaxFacts.FindTokenOnLeftOfPosition()` returned the first using keyword token in the file and that didn't have a parent (which should have been the using statement), but we weren't able to figure out how that's possible.  Then with @jasonmalinowski's help we learned that the crash occurred during typing when he was adding a named parameter to a call to `AsyncLazy::StartAsynchronousComputation` [here](https://github.com/dotnet/roslyn/blob/4b77f7647031df516cb6cf5d89364a3e812d2039/src/Workspaces/Core/Portable/Utilities/AsyncLazy%601.cs#L331), but a manual repro could not be found.  Given that we can at least prevent the crash by adding the conditional access operator, I'm going to merge in the changes but leave these notes here for future reference.

**Comment by brettfo** [comment]
test this please

**Comment by jasonmalinowski** [comment]
:+1:

**Comment by SLaks** [comment]
I have another repro of what looks like the same issue.

Can you please check?

Source: https://github.com/SLaks/DroidMaster/blob/roslyn-crash-1467/DroidMaster/UI/DeviceListViewModel_Commands.cs#L83

Place the cursor at the end of that line, open signature help, then press Ctrl+Home.

**Comment by brettfo** [comment]
@SLaks yes, this is the same issue.  Thanks for finding a manual repro for this!  I couldn't find one earlier so I was fixing this blindly.

**Pull request merged**